### PR TITLE
Improve Partition Identity performance

### PR DESCRIPTION
**Pull request by mhelleborg**
Rewrite of both HashRing and Partition Identity handover.

Breaking change with regards to Rendezvous algorithm for identity ownership, so can not run with older versions of Proto.Actor.

### CODE REVIEW

File: src/Proto.Cluster/Partition/MemberHashRing.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// -----------------------------------------------------------------------
+using System.Collections.Generic;
+using Proto.Router;
+
+namespace Proto.Cluster.Partition
+{
+    public class MemberHashRing : HashRing<Member>
+
+    {
+        public MemberHashRing(IReadOnlyCollection<Member> nodes) : base(nodes, member => member.Address, MurmurHash2.Hash, 50)
```
Comment (mhelleborg) [inline src/Proto.Cluster/Partition/MemberHashRing.cs:14]: Replicacount could perhaps be tuned, not sure what would be the best tradeoff here..

File: src/Proto.Cluster/Partition/PartitionIdentityRelocationWorker.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            Logger.LogError("Relocation timed out");
+            _onRelocationComplete.TrySetResult(false);
+            return Task.CompletedTask;
+        }
+
+        private Task OnDeadLetterResponse(DeadLetterResponse response, IContext context)
+        {
+            if (_waitingRequests.Remove(response.Target))
+            {
+                Logger.LogError("Unreachable node: {Address}", response.Target.Address);
```
Comment (mhelleborg) [inline src/Proto.Cluster/Partition/PartitionIdentityRelocationWorker.cs]: How is the best way to handle network partitions between members? If this is from a member which should still be in the cluster, we would be in an inconsistent state.

Comment (mhelleborg) [reply src/Proto.Cluster/Partition/PartitionIdentityRelocationWorker.cs]: More likely than network partition is that the member has already shut down though, which means that the node will need to rebalance again, and we will end up in a correct state.

File: src/Proto.Cluster/Partition/PartitionIdentityActor.cs [hunk 3]
```diff
@@ -93,40 +93,12 @@ private async Task OnClusterTopologyInner(ClusterTopology msg, IContext context)
 
             requestMsg.Members.AddRange(members);
```
Comment (cpx86) [inline src/Proto.Cluster/Partition/PartitionIdentityActor.cs]: A bit of a nit-pick, but this could be `msg.Members` and the `members` var could be scrapped, right?

Comment (mhelleborg) [reply src/Proto.Cluster/Partition/PartitionIdentityActor.cs]: Yes, would be a little cleaner. Fixing :)

File: src/Proto.Actor/Router/HashRing.cs [hunk 4]
```diff
@@ -9,33 +9,91 @@
...
+        }
+
+        public T GetNode(string key)
+        {
+            if (_ring.Length == 0) return default!;
+            if (_replicaCount == _ring.Length) return _ring[0].value;
+
+            var hash = _hash(key);
+
+            var result = Array.BinarySearch(_hashes, hash);
```
Comment (rogeralsing) [inline src/Proto.Actor/Router/HashRing.cs:48]: Nice!

File: src/Proto.Cluster/Partition/MemberHashRing.cs (hunk 1 above)
Comment (rogeralsing) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: My understanding of the stickiness is that the more replicas, the bigger the risk of not being sticky. e.g. more parts of the hash ring are moving.
So some config might be nice

File: src/Proto.Cluster/Partition/PartitionIdentityRelocationWorker.cs (hunk 2 above)
Comment (rogeralsing) [reply src/Proto.Cluster/Partition/PartitionIdentityRelocationWorker.cs]: Any member that become invisible to other members due to network partitioning, would be evicted from the cluster.
We have a fairly different approach than e.g. Akka which tries to "fix" things.
But in the era of let it crash, containers and pods.
Just let that node shut down, and restart

**Review by rogeralsing — COMMENTED** [review]
All in all, looks great. having config for hashring replicas would be nice though

File: src/Proto.Cluster/Partition/MemberHashRing.cs (hunk 1 above)
Comment (mhelleborg) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: The one potential problem with making it configurable is that two different configs can have different answers for which member an identity belongs to. If both configurations exists in a cluster, the behavior would be incorrect, and you could end up with lots of weird failure cases.  It might be slightly less error prone to choose a constant default value. 
The downside would be for someone who would like to tune the  distribution / stickyness. An option could be to make the partition identity check if the other nodes have the same settings as itself, and otherwise fail fast on startup.

Comment (cpx86) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: > make the partition identity check if the other nodes have the same settings

The same problem would occur if we were to change the constant replica count between versions, so maybe this is something we'd want to handle either way?

This got me thinking, is there any way to handle a change to the replica count gracefully? If all nodes must be in agreement on the replica count, then as long as there's one node alive then the value can't change. Meaning the only way to change it would be to shut down the whole cluster first, right?

Comment (mhelleborg) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: That is actually a very good point. The first node active in the cluster could be the master for replica count, and additional nodes would use that instead of any local settings. Then it can be tuned without worrying about it being out of sync across the cluster.

Comment (cpx86) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: A question then is what happens when the leader decides to change the replica count. Would the cluster become in sync quickly enough to avoid disruptions? Or would it be necessary to have some sort of coordinated pause until all nodes use the same value?

Comment (mhelleborg) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: I definitely don't think that needs to be dynamic at all. Keep the initial value from starting the cluster, no problem with updates.

Comment (rogeralsing) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: Is this a likely scenario though?
Under what conditions would a node decide that hey, let's tweak the replicas of the hash ring?

Comment (cpx86) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: Let me rephrase, because I botched my previous comment a bit 😅 

If 1) the replicaCount value changes (either because the const value changed between Proto.Cluster versions, or because it was changed via config) and 2) nodes always check what the current value in the cluster is (controlled by the oldest node), then wouldn't it mean the only way to change the value is to shut down all nodes? Even if you bring it down to a single-node cluster, that node would already be configured to use the old value. Or am I missing something?

Comment (mhelleborg) [reply src/Proto.Cluster/Partition/MemberHashRing.cs:14]: No, I think you summarized it :) The cluster will have to be shut down if you want to change replica count, if we go for this solution where we allow the first node to decide the setting.
It then later being removed from the cluster cannot change it, as that would break the behavior between the rest of the nodes.

**Comment by rogeralsing** [comment]
@mhelleborg , have this been tested with the cluster benchmark ?
otherwise I can give that part a go later

**Comment by mhelleborg** [comment]
> @mhelleborg , have this been tested with the cluster benchmark ? otherwise I can give that part a go later

Not yet. I do have  a few more changes I think should be included, but they should not affect performance. You can have a go testing it, but don't merge it yet :)

**Comment by mhelleborg** [comment]
In order to better handle compatibility between versions and potentially different configs, I think a handshake between nodes on startup would be helpful. Then they can verify that they are running with compatible configurations and within a compatible version range.  
Instead of running with potentially undefined behavior, it could instead stop before causing any problems.

File: src/Proto.Cluster/Gossip/Gossiper.cs [hunk 5]
```diff
@@ -97,9 +127,148 @@ private async Task GossipLoop()
             }
         }
 
+        public class ConsensusCheckBuilder<T>
```
Comment (mhelleborg) [inline src/Proto.Cluster/Gossip/Gossiper.cs]: Not married to the name..
